### PR TITLE
Add generic webhook support

### PR DESCRIPTION
**Pull request by LinuxSuRen**
## Why?

Normally, different SCM providers have their own webhook support. But not all SCM provides supported by Jenkins plugins.  So, it's quite important to have a generic webhook. In this PR, we take advantage of Jenkins [ generic webhook plugin](https://github.com/jenkinsci/generic-webhook-trigger-plugin).

Consider there're too many files in this PR. I'd suggest you review it with the following advice:

* please ignore those files which have comments like this `Code generated by controller-gen. DO NOT EDIT.`
  * all those files generated by `controller-gen`
* I changed the import path. We can belive it's ok if the compile process is ok.
  *  e.g. changes from `kubesphere.io/devops/api/v1alpha1` to `kubesphere.io/devops/pkg/api/devops/v1alpha1`
* CRD files were generated by `controller-gen` too. So, you can just ignore them
* Keep an eye on the following areas
  * I add some fields in the CRD `Pipeline`
  * I made some changes in the`Pipeline controller`
  * unit tests against the new functions

## How to use?

You can find the usage from the directory `docs`.

## Jenkins job config
Below is the snippet of a Jenkins job config.xml:
```
    <org.jenkinsci.plugins.workflow.job.properties.PipelineTriggersJobProperty>
      <triggers>
        <org.jenkinsci.plugins.gwt.GenericTrigger plugin="generic-webhook-trigger@1.72">
          <spec></spec>
          <regexpFilterText>$x_github_event</regexpFilterText>
          <regexpFilterExpression>push</regexpFilterExpression>
          <genericRequestVariables>
            <org.jenkinsci.plugins.gwt.GenericRequestVariable>
              <key>xxxx</key>
              <regexpFilter></regexpFilter>
            </org.jenkinsci.plugins.gwt.GenericRequestVariable>
          </genericRequestVariables>
          <genericHeaderVariables>
            <org.jenkinsci.plugins.gwt.GenericHeaderVariable>
              <key>X-GitHub-Event</key>
              <regexpFilter></regexpFilter>
            </org.jenkinsci.plugins.gwt.GenericHeaderVariable>
          </genericHeaderVariables>
          <printPostContent>true</printPostContent>
          <printContributedVariables>true</printContributedVariables>
          <causeString>Generic Cause - $name committed to $branch</causeString>
          <token>xxxx</token>
          <tokenCredentialId></tokenCredentialId>
          <silentResponse>false</silentResponse>
          <overrideQuietPeriod>false</overrideQuietPeriod>
        </org.jenkinsci.plugins.gwt.GenericTrigger>
      </triggers>
    </org.jenkinsci.plugins.workflow.job.properties.PipelineTriggersJobProperty>
```

Output image:

```
surenpi/devops-apiserver:dev--c7527cb
surenpi/devops-controller:dev--c7527cb
```

### CODE REVIEW

**Comment by LinuxSuRen** [comment]
hi @JohnNiang , please take a look at this or. Thanks.

File: README.md [hunk 1]
```diff
@@ -28,10 +32,17 @@ For example, you can access an API like:
 
 `curl http://ip:30880/kapis/clusters/host/devops.kubesphere.io/v1alpha3/devops/test847h4/credentials`
 
+## Code contribution
+
+If you're going to update or add CRD go struct, please run the following command once done with that:
+
+`make manifests generate generate-listers`
```
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply README.md:39]: I cannot tell the differences.

File: docs/webhook.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+HTTP request with some a generic payload. Users can use it from GitHub or just a curl command line. The backend rely on 
+[Jenkins generic-webhook-trigger plugin](https://github.com/jenkinsci/generic-webhook-trigger-plugin).
+
+For example, you can use the following command line to trigger some Pipelines:
+
+`curl -H "Content-Type: application/x-www-form-urlencoded" -X POST "http://ip:port/kapis/devops.kubesphere.io/v1alpha2/webhook/generic-trigger?token=xxxx"`
+
+You can get the output like below if you have a valid token:
+
+```
```
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply docs/webhook.md:26]: Is there any difference here?

File: hack/generate_group.sh [hunk 3]
```diff
@@ -76,6 +76,7 @@ fi
 if [ "${GENS}" = "all" ] || grep -qw "lister" <<<"${GENS}"; then
   echo "Generating listers for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/listers"
   ${GOPATH}/bin/lister-gen --input-dirs $(codegen::join , "${FQ_APIS[@]}") --output-package ${OUTPUT_PKG}/listers "$@"
+  echo ${GOPATH}/bin/lister-gen --input-dirs $(codegen::join , "${FQ_APIS[@]}") --output-package ${OUTPUT_PKG}/listers "$@"
```
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply hack/generate_group.sh:79]: Yes!

File: pkg/client/devops/jenkins/pipeline.go [hunk 4]
```diff
@@ -74,8 +74,10 @@ const (
 	CreateSCMServersUrl      = "/blue/rest/organizations/jenkins/scm/%s/servers/"
 	ValidateUrl              = "/blue/rest/organizations/jenkins/scm/%s/validate"
 
-	GetNotifyCommitUrl    = "/git/notifyCommit/?"
-	GithubWebhookUrl      = "/github-webhook/"
+	GetNotifyCommitUrl = "/git/notifyCommit/?"
+	GithubWebhookUrl   = "/github-webhook/"
+	// GenericWebhookUrl comes from Jenkins plugin, see also https://github.com/jenkinsci/generic-webhook-trigger-plugin
+	GenericWebhookUrl     = "/generic-webhook-trigger/invoke?"
```
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply pkg/client/devops/jenkins/pipeline.go:80]: I cannot find a reason to change other URLs.

[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply pkg/client/devops/jenkins/pipeline.go:80]: I cannot find a reason to change other URLs.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnNiang) [reply README.md:39]: hi @LinuxSuRen , it's my mistake, please have a look at previous comment again.

File: docs/webhook.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohnNiang) [reply docs/webhook.md:26]: The difference is [here](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks).

File: pkg/client/devops/jenkins/pipeline.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (JohnNiang) [reply pkg/client/devops/jenkins/pipeline.go:80]: IMO, base url should not end with question mark(?). Suppose we want to append a sub path which is very annoying.

File: docs/webhook.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+## Generic Webhook
+
+It does not require a particular payload in this kind of webhook. It accepts a standard 
+HTTP request with some a generic payload. Users can use it from GitHub or just a curl command line. The backend rely on 
+[Jenkins generic-webhook-trigger plugin](https://github.com/jenkinsci/generic-webhook-trigger-plugin).
+
+For example, you can use the following command line to trigger some Pipelines:
+
+`curl -H "Content-Type: application/x-www-form-urlencoded" -X POST "http://ip:port/kapis/devops.kubesphere.io/v1alpha2/webhook/generic-trigger?token=xxxx"`
```
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply docs/webhook.md:22]: As you can see from the following screenshot:

image

we cannot see the full command line with multi-line style and bash flag. And I cannot see many differences with or without the bash flag.

File: docs/webhook.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply docs/webhook.md:26]: I see. As you can see from the following screenshot:

image

There're not many differences.

File: pkg/client/devops/jenkins/pipeline.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply pkg/client/devops/jenkins/pipeline.go:80]: I got you. Considering the Jenkins client should be Refactored, I prefer to keep them in the same style for now.

File: docs/webhook.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohnNiang) [reply docs/webhook.md:26]: Apparently, the code blocks with syntax highlighting are easier to read than those without.

File: docs/webhook.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (JohnNiang) [reply docs/webhook.md:22]: The code blocks with syntax highlighting are easier to read than those without.

[the remark this answers is not in the document]
Comment (JohnNiang) [reply docs/webhook.md:22]: image

[the remark this answers is not in the document]
Comment (LinuxSuRen) [reply docs/webhook.md:22]: In most cases, people read this file from GitHub. And it's only a curl command line in this case. I prefer that we can see the whole command line without scrolling your mouse. I'd love to use the syntax flag if there're many shell script lines.

[the remark this answers is not in the document]
Comment (JohnNiang) [reply docs/webhook.md:22]: I still keep my opinion. But you don't have to make any changes here.

**Pull request merged**